src/Builder/Type/CombinedFieldQuery.php

```diff
@@ -4,28 +4,80 @@
 
 namespace MongoDB\Builder\Type;
 
-use MongoDB\BSON\Document;
-use MongoDB\BSON\Serializable;
 use MongoDB\Exception\InvalidArgumentException;
 use stdClass;
 
+use function array_is_list;
+use function array_key_exists;
+use function array_key_first;
+use function array_merge;
+use function array_reduce;
+use function count;
 use function get_debug_type;
+use function get_object_vars;
 use function is_array;
 use function sprintf;
+use function str_starts_with;
 
 /**
- * List of filters that apply to the same field path.
+ * List of field query that apply to the same field path.
  */
 class CombinedFieldQuery implements FieldQueryInterface
 {
-    public function __construct(
-        /** @var list<FieldQueryInterface|Serializable|array|stdClass> $fieldQueries */
-        public readonly array $fieldQueries,
-    ) {
-        foreach ($fieldQueries as $fieldQuery) {
-            if (! $fieldQuery instanceof FieldQueryInterface && ! $fieldQuery instanceof Serializable && ! is_array($fieldQuery) && ! $fieldQuery instanceof stdClass) {
-                throw new InvalidArgumentException(sprintf('Expected filters to be a list of %s, %s, array or stdClass, %s given.', FieldQueryInterface::class, Document::class, get_debug_type($fieldQuery)));
+    /** @var list<FieldQueryInterface|array|stdClass> $fieldQueries */
+    public readonly array $fieldQueries;
+
+    /** @param list<FieldQueryInterface|array|stdClass> $fieldQueries */
+    public function __construct(array $fieldQueries)
+    {
+        if (! array_is_list($fieldQueries)) {
+            throw new InvalidArgumentException('Expected filters to be a list, invalid array given.');
+        }
+
+        // Flatten nested CombinedFieldQuery
+        $this->fieldQueries = array_reduce($fieldQueries, static function (array $fieldQueries, mixed $fieldQuery): array {
```

Comment:
Using a `FieldQueryInterface|array|stdClass` type on the `$fieldQuery` parameter would change some `InvalidArgumentException` to type errors, which is appropriate in this case:
```suggestion
        $this->fieldQueries = array_reduce($fieldQueries, static function (array $fieldQueries, FieldQueryInterface|array|stdClass $fieldQuery): array {
```

Reply:
The type was actually a lot longer list.